subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,196 @@
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+static void discover_failed(struct bt_conn *conn, int err)
+{
+	BT_DBG("conn %p", (void *)conn);
+
+	ias_client_cb->discover(conn, err);
+}
+
+static void client_reset(struct bt_ias_client *ias_client)
+{
+	bt_conn_unref(ias_client->conn);
+	(void)memset(ias_client, 0, sizeof(*ias_client));
+}
+
+int bt_ias_alert_write(struct bt_conn *conn, enum bt_ias_alert_lvl alert_lvl)
+{
+	int err;
+	uint8_t lvl = alert_lvl;
+
+	if (conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_HANDLE);
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client alert level write failed: %d", err);
+		return err;
+	}
+
+	return err;
+}
+
+static uint8_t bt_ias_alert_lvl_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	struct bt_gatt_chrc *chrc = (struct bt_gatt_chrc *)attr->user_data;
+
+	if (attr == NULL) {
+		ias_client_cb->discover(conn, -BT_ATT_ERR_ATTRIBUTE_NOT_FOUND);
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_FOUND);
+	}
+
+	client_by_conn(conn)->write_handle = chrc->value_handle;
+	return 0;
+}
+
+static uint8_t bt_ias_prim_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	int err = 0;
+	struct bt_gatt_service_val *data;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	if (!attr) {
+		BT_DBG("Discover complete\n");
+		return BT_GATT_ITER_STOP;
+	}
+
+	data = attr->user_data;
+
+	client->discover.uuid = alert_lvl_uuid;
+	client->discover.start_handle = attr->handle + 1;
+	client->discover.end_handle = data->end_handle;
+	client->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	client->discover.func = bt_ias_alert_lvl_disc_cb;
+
+	err = bt_gatt_discover(conn, &client->discover);
+	if (err) {
+		BT_ERR("Characteristic discover failed\n");
+		goto fail;
+	}
+	ias_client_cb->discover(conn, err);
+
+	return BT_GATT_ITER_STOP;
+fail:
+	discover_failed(conn, err);
+	client_reset(client);
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_ias_discover(struct bt_conn *conn)
+{
+	int err;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	if (client->conn != NULL) {
+		client->conn = bt_conn_ref(conn);
+	}
+
+	if (ias_client_cb == NULL) {
+		BT_DBG("NULL callback pointer");
+		return -EINVAL;
+	}
+
+	client->discover.uuid = ias_uuid;
+	client->discover.func = bt_ias_prim_disc_cb;
+	client->discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	client->discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	client->discover.type = BT_GATT_DISCOVER_PRIMARY;
+
+	err = bt_gatt_discover(conn, &client->discover);
+	if (err < 0) {
```

Comment:
We should reject if there's already a discovery procedure going on for this client

Reply:
Done (added flag to `bt_ias_client` to check whether discovery is in progress)